documentation/high-level-introduction.md

```diff
@@ -0,0 +1,257 @@
+
+  - It is possible to invoke C# properties and functions from msbuild. 
+  - This is useful for a number of use cases, including string and path manipulation and math operations.
+  - See the [documentation](https://docs.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2019) for more details.
+
+## MSBuild Tasks
+- These are C# types defined in a library. 
+- The specifics of creating a new task library is outside the scope of this guide.
+- There are two common base classes for Tasks:
+  - `Task` - A generic “I want to do some work” base class with and `Execute ()` method to override
+  - `ToolTask` - A base class that makes invoking a single external tool significantly easier
+- Parameters are passed via public properties defined on the task
+
+```c#
+public abstract class MySpecialTask : Task 
+{
+    public string MyOptionalArg { get; set; }
+    
+    [Required]
+    public string MyRequiredArg { get; set; }
+}
+```
+
+  - Properties can be marked [Required] which will fail the build if given an empty string value
+  - Strings are the common type, but you can use types like bool and msbuild will convert for you
+  - ITaskItem[] are need to pass Items (lists) to \ from tasks when the task may need to modify them
+- They are later invoked with:
+
+```xml
+     <UsingTask TaskName="MyTasks.MySpecialTask" AssemblyFile="MyTasks.dll" />
+     
+    <Target Name="_MySpecialTask">
+                    <MySpecialTask
+                            MyOptionalArg="$(MyOptionalArg)"
+                            MyRequiredArg="$(MyRequiredArg)"
+                            >
+                    </MySpecialTask>
+    </Target>
+```
+- You can attach a C# debugger to the task, but that can be tricky on mono since there is no attach to process
+- `SessionId` is needed on most C# tasks to enable “build remoting” for Visual Studio Windows.
+
+## Execution Model
+
+A very inaccurate but useful way of thinking about how msbuild processes is to split it into three parts.
+
+- Loading
+  - Starting with the csproj load each file included into a giant buffer
+    - Think C preprocessor gloming files together
+  - Some default values will be set in the global property dictionary
+- Property Eval
+  - Starting from the top, evaluate every node
+    - Property and Item Group will set items in the global context
+    - UsingTask will load task assemblies and prepare to invoke them
+    - Targets will define then and their ordering\inputs\output
+      - **Their contents will not be executed yet**
+  - This will respect conditions, and be done top to bottom based upon include order
+- Target Eval
+  - Starting with the defined “DefaultTargets", which for our purposes is always “Build” determine a list of target to invoke and their ordering
+  - Invoke one at a time, comparing their Input\Output against file timestamps like make
+
+
+## "SDK style” project and nuget injection
+  - SDK style projects and nugets are beyond the scope of this document 
```

Comment:
Link? Also, it might be good to combine this and the next line to prevent me from wondering why you'd bring it up in the first place.